### PR TITLE
Rule component: Add a new API `/api/v1/alerts/:state` for pull alerts from Thanos Rule

### PR DESCRIPTION
**Pull request by wklken**
We want to use a `pull` monitoring system to replace the `alertmanager`, which needs to fetch the `firing` alerts from Thanos Rule. So, after the rule apis added by this issue https://github.com/improbable-eng/thanos/pull/851 from V0.3.2, we add one more API to fetch the alerts in specific `state`, like `firing`

## Changes

- `/api/v1/alerts/:state`

## Verification

- Tested query and rule components locally
- Tested in a Thanos cluster which monitoring a lot of k8s clusters

BTW, while the `AlertingRule.active` is not public, it's difficult to mock the `Alerts` and do the unittest here, still trying:)

### CODE REVIEW

File: pkg/rule/api/v1_test.go [hunk 1]
```diff
@@ -178,6 +178,15 @@ func testEndpoints(t *testing.T, api *API) {
...
 			},
 		},
+		{
+			endpoint: api.alerts,
+			response: &AlertDiscovery{Alerts: []*Alert{}},
+		},
+		{
+			endpoint: api.stateAlerts,
+			params: map[string]string{"state": "firing"},
+			response: &AlertDiscovery{Alerts: []*Alert{}},
```
Comment (povilasv) [inline pkg/rule/api/v1_test.go]: Could you add proper tests  mock some firing alerts and check that they are returned currectly

File: pkg/rule/api/v1.go [hunk 2]
```diff
@@ -128,6 +130,27 @@ func (api *API) alerts(r *http.Request) (interface{}, []error, *qapi.ApiError) {
 	return res, nil, nil
 }
 
+func (api *API) stateAlerts(r *http.Request) (interface{}, []error, *qapi.ApiError) {
```
Comment (povilasv) [inline pkg/rule/api/v1.go]: why are we returning `interface{}` here?

Comment (povilasv) [reply pkg/rule/api/v1.go]: Why are there `[]error` and `*api.ApiError` here? This needs to return proper data.

**Comment by brancz** [comment]
How is this different from querying the `ALERTS` time-series?

**Comment by povilasv** [comment]
@wklken does Prometheus / alertmanager have a similiar api? Couldn't find this exact api.

**Comment by wklken** [comment]
@povilasv 
- alertmanger has the api
- we want to `pull` the firing alerts from thanos-rule directly, without alertmanger been deployed.
-since `/alerts` is supported by rule-api now, it's reasonable to add one more API to filter the `firing` alerts

**Comment by povilasv** [comment]
could you link me to that api? I couldn't find the `:state` part.

**Comment by wklken** [comment]
https://github.com/prometheus/alertmanager/blob/master/api/v2/openapi.yaml#L130

`/alerts`, in query `active` default is `true`; 

In rule, `/alerts` will return all alerts, include `pending/firing`;

**Comment by povilasv** [comment]
I think we want to mimic Prometheus API, so I'm against introducing new API if it's not in Prometheus.

I would suggest doing this in Prometheus first and then we could port it here.

Thoughts about this @bwplotka ?

**Pull request closed without merging**